bolt-sidecar/src/api/commitments/firewall_recv.rs

```diff
@@ -0,0 +1,622 @@
+        // tx to a single receiver.
+        let (api_event_tx, api_events_rx) = mpsc::channel(self.urls.len() * 2);
+        let (ping_tx, ping_rx) = broadcast::channel::<()>(1);
+        let (shutdown_tx, shutdown_rx) = broadcast::channel::<()>(1);
+
+        // a task to send pings to open connections to the servers at regular intervals
+        tokio::spawn(async move {
+            let ping_interval = tokio::time::interval(PING_INTERVAL);
+            tokio::pin!(ping_interval);
+
+            loop {
+                ping_interval.tick().await;
+                if let Err(err) = ping_tx.send(()) {
+                    error!(?err, "internal error while sending ping task")
+                }
+            }
+        });
+
+        let signal = self.signal.take();
+        if let Some(signal) = signal {
+            tokio::spawn(async move {
+                signal.await;
+                if let Err(err) = shutdown_tx.send(()) {
+                    error!(?err, "failed to send shutdown signal");
+                }
+            });
+        }
+
+        for url in &self.urls {
+            let url = url.clone();
+            let api_events_tx = api_event_tx.clone();
+            let ping_rx = ping_rx.resubscribe();
+            let shutdown_rx = shutdown_rx.resubscribe();
+            tokio::spawn(async move {
+                handle_connection(url, api_events_tx, ping_rx, shutdown_rx).await;
+            });
+        }
+
+        api_events_rx
+    }
+
+    /// Returns the local addr the server is listening on (or configured with).
+    pub fn urls(&self) -> &[Url] {
+        self.urls.as_slice()
+    }
+}
+
+async fn handle_connection(
+    url: Url,
+    api_events_tx: mpsc::Sender<CommitmentEvent>,
+    ping_rx: broadcast::Receiver<()>,
+    shutdown_rx: broadcast::Receiver<()>,
+) {
+    // TODO: fill this
+    let ws_config = WebSocketConfig { ..Default::default() };
+
+    loop {
+        let should_retry =
+            match connect_async_with_config(url.to_string(), Some(ws_config), false).await {
+                Ok((stream, response)) => {
+                    info!(?url, ?response, "opened websocket connection");
+                    let (write_sink, read_stream) = stream.split();
+
+                    // For each opened connection, create a new commitment processor
+                    // able to handle incoming message requests.
+                    let commitment_request_processor = CommitmentRequestProcessor::new(
+                        url.clone(),
+                        api_events_tx.clone(),
+                        write_sink,
+                        read_stream,
+                        ping_rx.resubscribe(),
+                        shutdown_rx.resubscribe(),
+                    );
+                    // Run the commitment processor indefinitely, reconnecting on failure.
+                    commitment_request_processor.await
```

Comment:
Was confused as to why this would trigger a reconnect, but the future itself returns a bool indicating that. Nice